### ISSUE TITLE
v4.0.0 Database is locked for more than 00:01:00.

### ISSUE THREAD

**Issue by vladar107** [mbdavid/LiteDB#925]
We have next situation: 
When we can't push message in MQ we store it in LiteDB
Also we have the job, that take messages one by one from LitleDB and try to push them to MQ.

Putting method:

>         public void SaveMessage(MessageDb message)
>         {
>             var col = _db.GetCollection<MessageDb>();
>             col.Insert(message);
>         }

Once, our MQ service was down and we have a lot of massages, that we've put one by one in LitleDB
And when job has started for reading

>           public MessageDb GetFirstMessage()
>           {
>               var col = _db.GetCollection<MessageDb>();
>               return col.FindOne(Query.All());
>           }

We have next exception 

>    LiteDB.LiteException: Timeout. Database is locked for more than 00:01:00.
>    at LiteDB.FileHelper.TryExec(Action action, TimeSpan timeout)
>    at LiteDB.StreamExtensions.TryLock(FileStream stream, Int64 position, Int64 length, TimeSpan timeout)
>    at LiteDB.FileDiskService.Lock(LockState state, TimeSpan timeout)
>    at LiteDB.LockService.Read()
>    at LiteDB.LiteEngine.<Find>d__9.MoveNext()
>    at LiteDB.LiteEngine.<Find>d__13.MoveNext()
>    at LiteDB.LiteCollection`1.<Find>d__17.MoveNext()
>    at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source)

After that when the job try to delete message from LiteDB via method:

>         public void DeleteMessageById(Guid id)
>         {
>             var col = _db.GetCollection<MessageDb>();
>             col.Delete(msg => msg.Id == id);
>         }

We get the exception

>    LiteDB.LiteException: Timeout. Database is locked for more than 00:01:00.
>    at LiteDB.LockService.Write()
>    at LiteDB.LiteEngine.Transaction[T](String collection, Boolean addIfNotExists, Func`2 action)

Looks like write lock wasn't realesed
I've seen this issue #597, but it doesn't look like our case

Interesting fact: it was just once since November 2017, and gone after app restart

**Comment by ju2pom** [comment]
I have a similar issue in the following context (LiteDB 4.1.2):
 - A windows service access the database file in a process
 - At the same time a desktop application tries to access the same database file
=> `Exception thrown: 'LiteDB.LiteException' in LiteDB.dll
Timeout. Database is locked for more than 00:01:00.`

Do I have to handle locks myself? Or is it supposed to work?

**Comment by mbdavid** [comment]
Hi, are you keeping open or close your datafile in windows service? Datafile lock/unlock automatic with in each method call.

Can you describe what operations are done in windows service and desktop service (insert/update/deleting/quering)?

**Comment by ju2pom** [comment]
I was keeping the database opened when the service was running. I then modified the code to have it opened only when necessary and it's much better (I had to pay attention to lazy linq query evaluation too).
The main operation done in the service is an _InsertBulk_ of  about 30k documents. And in the application it's either FindAll or more restricted queries.

Anyway now that I know I should not keep the file locked it should be ok.
Thanks.
